c/json.c

```diff
@@ -2061,13 +2556,259 @@ JsonObject *jsonObjectProperty(JsonObject *object, char *propertyName, int *stat
+        if (!jsonObjectHasKey(baseObject,overrideProp->key)){
+          copyJson(builder,merged,overrideProp->key,overrideProp->value);
+        }
+        overrideProp = jsonObjectGetNextProperty(overrideProp);
+      }
+      return status;
+    } else {
+      copyJson(builder,parent,parentKey,overrides);
+      return JSON_MERGE_STATUS_SUCCESS;
+    }
+  } else if (jsonIsArray(base)){
+    JsonArray *baseArray = jsonAsArray(base);
+    if (jsonIsArray(overrides)){
+      JsonArray *overridesArray = jsonAsArray(overrides);
+      int arrayPolicy = merger->flags & JSON_MERGE_FLAG_ARRAY_POLICY_MASK;
+      Json *merged = jsonBuildArray(builder,parent,parentKey,&errorCode);
+      int sizeB = jsonArrayGetCount(baseArray);
+      int sizeO = jsonArrayGetCount(overridesArray);
+      int i;
+      switch (arrayPolicy){
+      case JSON_MERGE_FLAG_CONCATENATE_ARRAYS: /* len(merge) = len(a)+len(b) */
+        for (i=0; i<sizeB; i++){
+          copyJson(builder,merged,NULL,jsonArrayGetItem(baseArray,i));
+        }
+        for (i=0; i<sizeO; i++){
+          copyJson(builder,merged,NULL,jsonArrayGetItem(overridesArray,i));
+        }
+        return JSON_MERGE_STATUS_SUCCESS;
+      case JSON_MERGE_FLAG_MERGE_IN_PLACE:     /* len(merge) = max(len(a),len(b)), a overrides corresponding elements of b */
+        {
+          int mergedSize = (sizeB > sizeO ? sizeB : sizeO);
+          for (i=0; i<mergedSize; i++){
+            Json *valueB = (i < sizeB ? jsonArrayGetItem(baseArray,i) : NULL);
+            Json *valueO = (i < sizeO ? jsonArrayGetItem(overridesArray,i) : NULL);
+            if (valueO && valueB){
+              mergeJson1(merger,merged,NULL,valueO,valueB);
+            } else if (valueO){
+              copyJson(builder,merged,NULL,valueO);
+            } else if (valueB){
+              copyJson(builder,merged,NULL,valueB);
+            } else {
+              printf("*** Panic *** internal array merge error\n");
+            }
+          }
+        }
+        return JSON_MERGE_STATUS_SUCCESS;
+      case JSON_MERGE_FLAG_TAKE_BASE:          /* len(merge) = len(b) */
+        for (i=0; i<sizeB; i++){
+          copyJson(builder,merged,NULL,jsonArrayGetItem(baseArray,i));
+        }
+        return JSON_MERGE_STATUS_SUCCESS;
+      case JSON_MERGE_FLAG_TAKE_OVERRIDES:     /* len(merge) = len(a) */
+      default: 
+        for (i=0; i<sizeO; i++){
+          copyJson(builder,merged,NULL,jsonArrayGetItem(overridesArray,i));
+        }
+        return JSON_MERGE_STATUS_SUCCESS;
+      }
+    } else {
+      copyJson(builder,parent,parentKey,overrides);
+      return JSON_MERGE_STATUS_SUCCESS;
+    }
+  } else {
+    /* printf("merge scalar case parentKey=%s\n",(parentKey ? parentKey : "<noKey>"));
+       fflush(stdout);
+    */
+    copyJson(builder,parent,parentKey,overrides);
+    return JSON_MERGE_STATUS_SUCCESS;
+  }
+}
+
+Json *jsonMerge(ShortLivedHeap *slh, Json *overrides, Json *base, int flags, int *statusPtr){
```

Comment:
If `overrides` and `base` are never changed, making them `const` could be a good way to tell users they're not going to be modified.

Reply:
I need a "general improvements" some day.